src/lib/util/Config.ts

```diff
@@ -110,12 +113,35 @@ export class Config {
 
 		this.typeChecker = new TypeChecker();
 		this.fileHandler = new FileHandler();
+		this.versionUpgradeManager = new VersionUpgradeManager();
 		this.sfdxScannerPath = Controller.getSfdxScannerPath();
 		this.configFilePath = path.join(this.sfdxScannerPath, CONFIG_FILE);
 		this.logger.setLevel(LoggerLevel.TRACE);
 		await this.initializeConfig();
 
 		this.initialized = true;
+		// Upgrades MUST happen AFTER the config is marked as initialized, to prevent any unexpected recursion.
+		await this.upgradeConfig();
+	}
+
+	private async upgradeConfig(): Promise<void> {
+		if (this.versionUpgradeManager.upgradeRequired(this.configContent.currentVersion)) {
+			let error: Error = null;
+			try {
+				this.configContent.currentVersion = await this.versionUpgradeManager.upgradeToLatest(this.configContent, this.configContent.currentVersion);
```

Comment:
Should we save the existing config to a backup file, maybe `Config.json.<old_version>.bak` before performing the upgrade?

Reply:
I suppose we could, but the changes aren't actually persisted until after all of the scripts run, so it's possibly unnecessary.

Reply:
How about preserving it and writing a backup only if there is an error. We want to avoid accidentally losing any changes someone might have done by hand. If we detect an error there seems like a higher probability that we might lose someone's manual edits.

Reply:
Good point. I'll do that.